doc/libpmemobj/pmemobj_tx_begin.3.md

```diff
@@ -315,6 +323,80 @@ The **TX_END** macro cleans up and closes the transaction started by the
 It is mandatory to terminate each transaction with this macro. If the transaction
 was aborted, *errno* is set appropriately.
 
+## TRANSACTION LOG TUNING ##
+
+From pmemobj implementation perspective there are two types of operations
+in a transaction:
+
++ **snapshots**, where action must be persisted immediately,
++ **intents**, where action can be persisted at the transaction commit phase
+
+**pmemobj_tx_add_range**(3) and all its variants belong to the **snapshots**
+group.
+
+**pmemobj_tx_alloc**(3) (with its variants), **pmemobj_tx_free**(3),
+**pmemobj_tx_realloc**(3) (with its variants) and **pmemobj_tx_publish**(3)
+belong to the **intents** group. Even though **pmemobj_tx_alloc**() allocates
+memory immediately, it modifies only the runtime state and postpones persistent
+memory modifications to the commit phase. **pmemobj_tx_free**(3) cannot free
+the object immediately, because of possible transaction rollback, so it
+postpones both the action and persistent memory modifications to the commit
+phase. **pmemobj_tx_realloc**(3) is just a combination of those two.
+**pmemobj_tx_publish**(3) postpones reservations and deferred frees to
+the commit phase.
+
+Those two types of operations (snapshots and intents) require that pmemobj
+builds a persistent log of operations. Intent log (also known as a "redo log")
+is applied on commit and snapshot log (also known as an "undo log")
+is applied on abort.
+
+When pmemobj transactions starts, it's not possible to predict how much
+persistent memory space will be needed for those logs. This means that pmemobj
+must internally allocate this space whenever it's needed. This has two downsides:
+
++ when transaction snapshots a lot of memory or does a lot of allocations,
+  pmemobj may need to do many internal allocations, which must be freed when
+  transaction ends, adding a lot of time overhead when such big transactions are
+  frequent,
++ when pool is almost full, freeing in a transaction may actually fail because of
```

Comment:
`transactions can start to fail due to not enough log space. This can be especially problematic for transactions that want to deallocate objects, as those might also fail. `